node-manager/src/encrypt.rs

```diff
@@ -0,0 +1,82 @@
+use aes_gcm::aead::Aead;
+use aes_gcm::{Aes256Gcm, KeyInit, Nonce};
+use pbkdf2::password_hash::Output;
+use pbkdf2::password_hash::{PasswordHasher, Salt, SaltString};
+use pbkdf2::{Params, Pbkdf2};
+use rand_core::{OsRng, RngCore};
+
+// Copied from https://github.com/FAE56/wasm-encrypt-rs/blob/master/src/crypto.rs
+
+pub fn encrypt(content: &str, password: &str) -> String {
+    let mut salt = [0u8; 16];
+    OsRng.fill_bytes(&mut salt);
+    let derive_key = derive_key(password, &salt);
+    let key = derive_key.as_bytes();
+
+    let mut iv = [0u8; 12];
+    OsRng.fill_bytes(&mut iv);
+
+    let cipher = Aes256Gcm::new_from_slice(key).unwrap();
+    let nonce = Nonce::from_slice(&iv);
+    let mut bytes = cipher.encrypt(nonce, content.as_bytes()).unwrap();
+
+    let mut combined = vec![];
+    combined.append(&mut salt.to_vec());
+    combined.append(&mut iv.to_vec());
+    combined.append(&mut bytes);
+    base64::encode(combined.as_slice())
+}
+
+pub fn decrypt(encrypted: &str, password: &str) -> String {
+    let buffer = base64::decode(encrypted).expect("Error reading ciphertext");
+    let buffer_slice = buffer.as_slice();
+    let salt = &buffer_slice[0..16];
+    let iv = &buffer_slice[16..28];
+    let data = &buffer_slice[28..];
+
+    let derive_key = derive_key(password, salt);
+    let key = derive_key.as_bytes();
+
+    let cipher = Aes256Gcm::new_from_slice(key).unwrap();
+    let nonce = Nonce::from_slice(iv);
+    let decrypted = cipher.decrypt(nonce, data).unwrap();
+    String::from_utf8(decrypted).unwrap()
+}
+
+fn derive_key(password: &str, salt: &[u8]) -> Output {
+    let params = Params {
+        rounds: 12345,
+        output_length: 32,
+    };
```

Comment:
It would be good to confirm these parameters, because these can't change later. I had to go through something similar before because we didn't have enough rounds. And too many rounds cause memory issues. Need to look at how many to do based on security but also practicality.

Reply:
Yeah need to be conscious of mobile phones here too

Reply:
tried this on my phone and definitely noticed a slow down. I think 2048 will be a good number to pick here, won't be _as secure_ but we are using BIP 39 which uses this anyways so we are falling back to that level of security anyway. Tested 2048 on my phone and didn't notice any real slow down

Reply:
Something we could also add is a version number to our database so we can migrate if needed

Reply:
Here's an interesting article about how lastpass went about it: https://support.1password.com/pbkdf2/

Reply:
But yeah this is good to merge and we can think about it later too